Remove CocoaPods

mokagio opened the pull request:
This is a replay of #23951 but onto https://github.com/wordpress-mobile/WordPress-iOS/pull/23923, as discussed with @kean [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/23951#issuecomment-2576774459). I copied the PR description below, but you might want to check the original PR, too, for the inline comments.

---

Notice the base is #23960, which addresses the build failures from #23923, so that we can verify that CI is green with only the CocoaPods changes. Refer to that PR for explanations for UI tests failures, too.

---

Picks up where #23750, but starting from scratch because of too many project file changes since then, and completes the migration away from CocoaPods as the mean of linking the Gutenberg XCFramework.

The sequence of changes is roughly:

- De-integrate CocoaPods
- Drag-n-drop the `xcframework`s
- Tweak build settings to work around compilation errors due to the frameworks see fbb0997a21089d01890222fa8568c9e67946c49c and 4a29aeacd770482bf998d01aa3bea74a09cfcf15
- Update the script that copies the React Native `App.js` file from the XCFramework to the build products as a `jsbundle` to use the new location
- Remove the various "if cocoapods" kind of checks from app code and automation

Theree things noteworthy:

- I had to disable a couple of clang checks, `CLANG_WARN_STRICT_PROTOTYPES` and `CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER`, because of build failures in the XCFramework imports. It doesn't seem ideal, but also we have increasingly less Objective-C so it should be acceptable. Open to suggestions.
- @jkmassel 's implementation had a build phase that downloaded the XCFrameworks sources. I omitted it and **expect devs to run `make dependencies` before building the app**. It felt a build phase was a bit too much, but it also feels like a line in the `README` to call attention to this is not enough. Open to suggestions.
- @jkmassel 's implementation also deleted the `xcworkspace` in favor of using only the `xcproject`, which makes sense given the workspace was required by the CocoaPods setup which is now gone. I decided not to remove the workspace in this PR just to not add to an already big diff. Also, when I opened the standalone project file, some things weren't working out of the box, so I decide to postpone the move to it.

### SwiftLint

SwiftLint was previously fetched via CocoaPods and the automation was configured with that in mind. Removing CocoaPods also removes SwiftLint. In the interest of moving on with the removal process, _I didn't provide an alternative setup for SwiftLint._ The `rake lint` task merely prints a "no linter configured" message at the moment. As you can see in CI, this does not affect SwiftLint there, so we still have reliable checks in place.

## Testing

I published this from the Simulator: https://giotestdotsite.wordpress.com/2025/01/07/via-gutenbegkit-without-cocoapods/ (please ignore the slug, it's indeed mobile-gunteber)   — Not sure if the HTML source can show it, but I did check locally that the experimental editor toggle was disabled. Besides, the UI being a bit different between the two helps being aware of it.

Apart from that, I'd say run the prototype build on device and see if the Gutenber-via-XCFramework editor works as expected.

## Regression Notes
1. Potential unintended areas of impact

It's possible that some behavior in the Gutenber-XCFramework editor will be compromised. However, I don't see why, once the editor loads, all the code should be the same.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Had a go at publishing a post from the Simulator. 

3. What automated tests I added (or what prevented me from doing so)

N.A.

---

PR submission checklist:

- [x] I have completed the Regression Notes. 
- [x] I have considered adding unit tests for my changes. - N.A.
- [x] I have considered adding accessibility improvements for my changes. - N.A.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. - N.A.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

mokagio committed:
Run `pod deintegrate` and remove CocoaPods automation

mokagio committed:
Add `Makefile` automation to download Gutenberg framework

mokagio committed:
Remove dead reference to `Pods.framework` in project

mokagio committed:
Remove dead reference to pods from workspace

mokagio committed:
Remove SwiftLint-via-CocoaPods from `Rakefile`

mokagio committed:
Remove unnecessary `$(inherited)` flag from a couple of targets

That's for the value `CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER`,
but it's useless to set `$(inherited)` for a Yes or No value because
removing the value from the target will automatically make Xcode read it
from the project, achieving the same result.

mokagio committed:
Remove CocoaPods instruction from the README

mokagio committed:
Remove SwiftLint instructions from the `README`

mokagio committed:
Add a version of Aztec fetched via SwiftPM

mokagio committed:
Add Gutenberg XCFrameworks via drag-n-drop from local paths

mokagio committed:
Disable `CLANG_WARN_QUOTED_INCLUDE_IN_FRAMEWORK_HEADER`

Was giving errors in Yoga:

> Double-quoted include "YGEnums.h" in framework header, expected angle-bracketed instead

mokagio committed:
Disable `CLANG_WARN_STRICT_PROTOTYPES`

It resulted in a build failure:

> React.framework/Headers/RCTAppearance.h:16:60:
> A function declaration without a prototype is deprecated in all versions of C

mokagio committed:
Update search path for Gutenberg React Native bundle in script

mokagio committed:
Remove some leftover bits of CocoaPods set up for automation

mokagio committed:
Remove leftover `COCOAPODS` precompiler flag

Strange that `pod deintegrate` did not remove it. Maybe it was some
custom piece of setup...

mokagio committed:
Remove a couple of CocoaPods-specific code branches

mokagio committed:
Temporarily switch to CI toolkit branch compatible with no CocoaPods

mokagio committed:
Fetch XCFrameworks in CI

mokagio committed:
Update Gutenberg checkout logic in localization lint for CP-less setup

mokagio committed:
Update source map upload for CP-less Gutenberg setup

mokagio committed:
Ensure `Frameworks` folder exist, otherwise tar will fail

mokagio committed:
Remove "Log Outdated Pods" step from code freeze

We don't have pods anymore

mokagio committed:
Remove Pods from lists in `.gitignore` and RuboCop

We don't use Pods anymore

mokagio committed:
Remove imports of Pods xcconfigs

mokagio committed:
No longer look in Pods folder to extract localizations

mokagio committed:
Link Gutenberg XCFramework in Jetpack target too!

mokagio committed:
Add instruction in README to run `make dependencies`

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Use CI toolkit 3.8.0

kean commented:
Hey, nice work. It's a historic moment 😃 

I've updated the tests in the target branch. It should be green now. I don't think I should/can review the changes in infra, but you get 👍

kean commented on Podfile:
```diff
@@ -1,126 +0,0 @@
-end
-
-abstract_target 'Tools' do
-  pod 'SwiftLint', swiftlint_version
```
Where does the SwiftLint version get pinned now? It doesn't appear to be in Gemfile. I'd probably suggest trying SwiftPM since this project is not itself a dependency, so it should be fine.

mokagio replied:
> Where does the SwiftLint version get pinned now?

The version is defined in `.swiftlint.yml`:

https://github.com/wordpress-mobile/WordPress-iOS/blob/388dbb167b7ae6fd763ed4c90076dc94679a9dc6/.swiftlint.yml#L1

As for installing the tool, I'm happy to have a go at setting it up via SwiftPM. I ignored the problem for the context of this PR because there was already a lot going on.

I was also keen to ask everyone in the team what they'd prefer to do for provisioning. @jkmassel and I discussed. in the past the idea of having Apps Infra recommend and maintain a golden path for integration (which was CocoaPods but now needs to be updated) but leave the teams the option to decide how to go about it. _For example, some people/team in the past expressed the feeling that running the linter on every build was too much friction when just hammering code together and they'd rather get linter feedback in the PR via CI only._

kean replied:
I'd suggest removing `swiftlint_version` from `.swiftlint.yml` and pinning it using SwiftPM. It should solve both issues: installation and pinning. Wdyt?

Note: I never tried it. I usually use Gemfile for tools.

> rather get linter feedback in the PR via CI only

I'm sorry, I don't believe anyone would've preferred this 😄

mokagio replied:
> I'd suggest removing `swiftlint_version` from `.swiftlint.yml` and pinning it using SwiftPM. It should solve both issues: installation and pinning. Wdyt?

Sounds good, but... currently the linter expects `swiftlint_version` to be defined in the config file, see https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/29064ab1e6e49135f9700d86b7b3279081b9ebc9/bin/run_swiftlint#L25

I think we could satisfy both the SPM and `swiftlint_version` requirements by reading the SwiftLint version from the YML in the Package file, see https://github.com/wordpress-mobile/WordPress-iOS-Shared/blob/6e8a44120452b08075d1838506bd632a5f83b044/Package.swift#L55-L77

In the past, that approach has given us trouble when implemented in a dependency, but given this is a root project, we should be fine. I'll try it in a dedicated PR.

mokagio replied:
~~Just a note that I had a look at this and is not as straightforward as I hoped.~~

~~@kean @crazytonyli @jkmassel would you be okay with merging without the SwiftLint support, developing for a few days without the linter?~~

Update: Still not super straightforward, and with some performance issues. But I have a SwiftPM implementation that ticks most boxes here https://github.com/wordpress-mobile/WordPress-iOS/pull/23996

~~If you take a look, you'll see it runs slower than 0.54.0. Depending on the performance consideration, we might consider sticking with CocoaPods just to fetch SwiftLint—at least till SwiftLint addresses the issue. Given it would be used only for a CLI tool outside the project, I'd consider it a minor DX annoyance.~~ Update: I looked better into the numbers and I can say the performance is comparable to what's on `trunk` with the existing setup. The poor execution time I saw originally on my end was due to configuration issues.

mokagio commented on WordPress/Makefile:
```diff
@@ -0,0 +1,10 @@
+	tar -xvf Frameworks/Gutenberg.tar.gz -C Frameworks/ -k
+	mv -n Frameworks/Frameworks/*.xcframework Frameworks/
+	rm -rf Frameworks/Frameworks Frameworks/dummy.txt
+	mkdir -p Frameworks/hermes.xcframework/ios-arm64/dSYMs Frameworks/hermes.xcframework/ios-arm64_x86_64-simulator/dSYMs
```
Given `gutenberg-mobile` should have an hopefully short time left in the project, I think this rudimental way to fetch it is acceptable.

Also, the version is not likely to change at all, which again makes me comfortable with simply hardcoding it there and duplicating it it in Fastlane.

crazytonyli replied:
Is it possible to use `binaryTarget` to import gutenberg-mobile via SPM?

crazytonyli replied:
If not, can we use the existing Rakefile, instead of adding a new Makefile?

mokagio replied:
> Is it possible to use `binaryTarget` to import gutenberg-mobile via SPM?

I haven't tried using SwiftPM directly yet. I inherited this setup from #23750 and it worked 😄 How about trying it as a followup?

> If not, can we use the existing Rakefile, instead of adding a new Makefile?

Yeah, that's a good point. I'll take care of it.

mokagio commented on WordPressAuthenticator/Sources/Services/LoginFacade.m:
```diff
@@ -1,6 +1,6 @@
 #import "LoginFacade.h"
 #import "WordPressXMLRPCAPIFacade.h"
-#import "WPAuthenticator-Swift.h"
+#import <WordPressAuthenticator/WordPressAuthenticator-Swift.h>
```
The header above was necessary when the lib was distributed via CocoaPods, but now we can access the Xcode-generated one directly.

This is unrelated with the CP removal for Gutenberg, but I noticed it when grepping "CocoaPods".

twstokes commented on Modules/Package.swift:
```diff
@@ -50,6 +50,7 @@ let package = Package(
         .package(url: "https://github.com/Automattic/wordpress-rs", revision: "alpha-20241116"),
         .package(url: "https://github.com/wordpress-mobile/GutenbergKit", revision: "4476d597b271778d001a26c50d74e527b54ebfef"),
         .package(url: "https://github.com/Automattic/color-studio", branch: "trunk"),
+        .package(url: "https://github.com/allenhumphreys/AztecEditor-iOS", branch: "ah/spm"),
```
👋 @mokagio - do we want to point Aztec to a fork? If I understand correctly this will block us if we need to push out the rare Aztec fix. Or maybe there's a timeline to get back to the WP org repo shortly after this change?

mokagio replied:
Agreed. This was in the PR that I based my work on, and then I forgot to follow up.

I'll get started porting the changes from that fork into the official repo.

twstokes replied:
Sounds good!

mokagio replied:
Closing the loop: https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1406

kean commented:
Hey, we've merged the [feature branch](https://github.com/wordpress-mobile/WordPress-iOS/pull/23923) in `trunk`, so please feel free to continue the work branching from `trunk`.

mokagio committed:
Point Aztec to branch in wordpress-mobile org from SPM support PR

See https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1379

mokagio committed:
Point Aztec to latest SwiftPM-support commit

crazytonyli commented:
Hey @mokagio , there are whole lots of unrelated commits in this PR. You may want to remove them before merging this PR?

crazytonyli commented on Modules/Package.swift:
```diff
@@ -50,6 +50,7 @@ let package = Package(
         .package(url: "https://github.com/Automattic/wordpress-rs", revision: "alpha-20241116"),
         .package(url: "https://github.com/wordpress-mobile/GutenbergKit", revision: "cc52214a50893b41898607ac0bff7f2787b085bb"),
         .package(url: "https://github.com/Automattic/color-studio", branch: "trunk"),
+        .package(url: "https://github.com/wordpress-mobile/AztecEditor-iOS", revision: "ea5f9aa13e7490647c8ff5c0aac7f9c25614a8df"),
```
Change commit to a releas tag?

mokagio replied:
Released https://github.com/wordpress-mobile/AztecEditor-iOS/releases/tag/1.20.0 and updated in [1780f9b](https://github.com/wordpress-mobile/WordPress-iOS/pull/23958/commits/1780f9bc60d635f2a4ea39e07d8622dc732be707)

mokagio commented:
Thanks for the review @crazytonyli , brings back good memories.

>  there are whole lots of unrelated commits in this PR. You may want to remove them before merging this PR?

Oh those must be a byproduct of this build being based off https://github.com/wordpress-mobile/WordPress-iOS/pull/23923 and it having been "Squash and merge"ed.

I'll rebase.

mokagio committed:
Point to latest Aztec, 1.20.0

mokagio committed:
Reimplement logic to download XCFrameworks from `make` to `rake`

mokagio commented:
@crazytonyli 

> If not, can we use the existing Rakefile, instead of adding a new Makefile?

Done in [d7d0010](https://github.com/wordpress-mobile/WordPress-iOS/pull/23958/commits/d7d00100df6b7717ed6efbf0193d2d120e2d5460)

crazytonyli commented on Scripts/BuildPhases/CopyGutenbergJS.sh:
```diff
@@ -3,36 +3,29 @@
 BUNDLE_ASSETS="$DEST/assets/"
 
 if [[ -d $XCFRAMEWORK_BUNDLE_ROOT ]]; then
+  set +x
```
Should this line be removed?

crazytonyli commented on Rakefile:
```diff
@@ -103,43 +102,25 @@ bundle exec fastlane run configure_apply force:true
+    sh "curl https://cdn.a8c-ci.services/gutenberg-mobile/Gutenberg-#{GUTENBERG_VERSION}.tar.gz --output #{gutenberg_tar_gz_download_path} -C -"
+    sh "tar -xf #{gutenberg_tar_gz_download_path} -C #{frameworks_dir}/ -k"
+    sh "mv -n #{frameworks_dir}/Frameworks/*.xcframework #{frameworks_dir}/"
+    sh "rm -rf #{frameworks_dir}/Frameworks #{frameworks_dir}/dummy.txt"
```
Nitpick: FileUtils has API for mkdir, mv, and rm commands.

mokagio replied:
Worked on it through 1590586 and 2d542d1. Pushed it forward too, re-implementing the download and unarchiving in Ruby (with Cursor's help)

crazytonyli replied:
Thanks for looking into it! It'd be nice if Rake has a feature to download a file (with checksum would be even nicer) and unzip a file.

mokagio committed:
Remove a debug `set +x`

mokagio committed:
Rewrite `gutenberg_xcframeworks` using Ruby APIs

More idiomatic and safer than shelling out. See
https://github.com/wordpress-mobile/WordPress-iOS/pull/23958#discussion_r1942316307

mokagio committed:
Also delete `tar.gz` when cleaning up after setting up Gutenberg

mokagio committed:
Ignore `Pods/` for retro-compatibility

crazytonyli reviewed (approved):
Thanks for working on this! And, bye-bye CocoaPods 🥳
